File: 프로그래머스/3/64062. 징검다리 건너기/징검다리 건너기.cpp
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

vector<int> cStones;
int jump;

bool is_possible(int num) {
    int tmpJump = 0;
    for(int i = 0; i < cStones.size(); i++) {
        if(cStones[i] < num) {
            tmpJump++;
            if(tmpJump >= jump) return false;
        }
        else {
            tmpJump = 0;
        }
    }
    return true;
}

int solution(vector<int> stones, int k) {
    int answer = 0;
    cStones = stones;
    jump = k;
    
    int left = 1, right = *max_element(stones.begin(), stones.end());
    
    while(left <= right) {
        int mid = (left + right) / 2;
        
        if(is_possible(mid)) {
            answer = mid;
            left = mid + 1;
        }
        else {
            right = mid - 1;
        }
    }
    
    return answer;
}

File: 프로그래머스/3/64062. 징검다리 건너기/징검다리 건너기.py
def is_possible(stones: list, friends: int, max_jump: int) -> bool:
    cur_jump = 1
    for stone in stones:
        if cur_jump > max_jump:
            return False
        
        if stone < friends:
            cur_jump += 1
        else:
            cur_jump = 1
        
    if cur_jump > max_jump:
        return False
    return True

def solution(stones, k):
    answer = 0
    
    start, end = 0, max(stones)
    
    while start <= end:
        mid = (start + end) // 2
        
        if is_possible(stones, friends=mid, max_jump=k):
            start = mid + 1
            answer = mid
        else:
            end = mid - 1
    
    return answer